lib/Conversion/StandardToHandshake/StandardToHandshake.cpp

```diff
@@ -631,6 +644,421 @@ int getBranchCount(Value val, Block *block) {
   return uses;
 }
 
+static bool hasMultiplePredecessors(Block *block) {
+  return std::distance(block->getPredecessors().begin(),
+                       block->getPredecessors().end()) > 1;
+}
+
+// The BFS callback provides the current block as an argument and returns
+// whether search should halt.
+enum class BFSNextState { Halt, SkipSuccessors, Continue };
+using BFSCallback = llvm::function_ref<BFSNextState(Block *)>;
+
+static void blockBFS(Block *start, BFSCallback callback) {
+  std::set<Block *> visited;
```

Comment:
It might be worth switching to the LLVM data structures for this. Would a DenseSet and SmallVector work for `visited` and `queue`?